### ISSUE TITLE
Fallback translations can lead to infinite recursion and SEGFAULTS

### ISSUE THREAD

**Issue by nussjustin-hmmh** [wirecard/magento2-ee#225]
### Prerequisites

- [X] I have my [server set up correctly](https://github.com/wirecard/magento2-ee/wiki/Installation#requirements).
- [ ] I am using a [supported version of Magento](https://github.com/wirecard/magento2-ee/wiki#overview-magento-2-shop-plugin).

### Versions

**Magento version:** 2.3.1
**PHP version:** 7.2.17

### Issue

**Problem description:**

The `magento2-ee` module wraps around the `\Magento\Framework\Phrase\Renderer\Translate::render` method using an `around` plugin method and tries to load fallback translations for the module, even if not needed.

Inside the `around` wrapper, the method [`\Magento\Framework\Phrase\Renderer\Translate::$translator->loadData`](https://github.com/wirecard/magento2-ee/blob/b2661caebc285c31957508764b2e6834d9e23660/Model/TranslationFallback.php#L63) method is called.

When Magento tries to convert a `Magento\Framework\Phrase` to a string (using `Magento\Framework\Phrase-::__toString`) while inside `loadData` or one of it's callees, the code can end up calling the `aroundRender` method of the `TranslationFallback` class again, which calls `loadData` agains which leads to infinite recursion and a SEGFAULT (or an exception, when Xdebug is installed).

This can happen for different reasons, e.g. when getting an `Area code not set` or `Required parameter 'theme_dir' was not passed` exception. In our case the code works perfectly without the `magento2-ee` module (since Magento 2.1, up to 2.3.1), but even with some workarounds (emulating area codes) we still have cases where our code ends up in this infinite loop.

From what I saw when looking at the code in the `TranslationFallback.php`, simply not loading the fallback translations should fix the problem in our case.

Part of a backtrace (generated using Xdebug, which throws an exception after 256 levels of nested calls) (area code is `adminhtml`):

```
PHP  23. Magento\Shipping\Model\ShipmentNotifier->notify()
PHP  24. Magento\Sales\Model\Order\Email\Sender\ShipmentSender->send()
PHP  25. Magento\Sales\Model\Order\Shipment->getOrder()
PHP  26. Magento\Sales\Model\OrderRepository\Interceptor->get()
PHP  27. Magento\Sales\Model\OrderRepository\Interceptor->___callPlugins()
PHP  28. Magento\Sales\Model\OrderRepository\Interceptor->Magento\Framework\Interception\{closure:vendor/magento/framework/Interception/Interceptor.php:104-151}()
PHP  29. Magento\GiftMessage\Model\Plugin\OrderGet->afterGet()
PHP  30. Magento\GiftMessage\Model\Plugin\OrderGet->getOrderGiftMessage()
PHP  31. Magento\GiftMessage\Model\OrderRepository->get()
PHP  32. Magento\Framework\Exception\NoSuchEntityException->__construct()
PHP  33. Magento\Framework\Exception\NoSuchEntityException->__construct()
PHP  34. Magento\Framework\Phrase->render()
PHP  35. Magento\Framework\Phrase\Renderer\Composite->render()
PHP  36. Magento\Framework\Phrase\Renderer\Translate\Interceptor->render()
PHP  37. Magento\Framework\Phrase\Renderer\Translate\Interceptor->___callPlugins()
PHP  38. Magento\Framework\Phrase\Renderer\Translate\Interceptor->Magento\Framework\Interception\{closure:vendor/magento/framework/Interception/Interceptor.php:104-151}()
PHP  39. Wirecard\ElasticEngine\Model\TranslationFallback\Interceptor->aroundRender()
PHP  40. Magento\Framework\Translate->loadData()
PHP  41. Magento\Framework\Translate->setConfig()
PHP  42. Magento\Theme\Model\View\Design\Proxy->getDesignTheme()
PHP  43. Magento\Theme\Model\View\Design->getDesignTheme()
PHP  44. Magento\Theme\Model\ThemeFactory->create()
PHP  45. Magento\Framework\App\ObjectManager->create()
PHP  46. Magento\Framework\ObjectManager\Factory\Compiled->create()
PHP  47. Magento\Framework\ObjectManager\Factory\Compiled->get()
PHP  48. Magento\Framework\ObjectManager\Factory\Compiled->create()
PHP  49. Magento\Framework\View\Design\Theme\Validator->__construct()
PHP  50. Magento\Framework\View\Design\Theme\Validator->_setTypeValidators()
PHP  51. Magento\Framework\Phrase->__toString()
PHP  52. Magento\Framework\Phrase->render()
PHP  53. Magento\Framework\Phrase\Renderer\Composite->render()
PHP  54. Magento\Framework\Phrase\Renderer\Translate\Interceptor->render()
PHP  55. Magento\Framework\Phrase\Renderer\Translate\Interceptor->___callPlugins()
PHP  56. Magento\Framework\Phrase\Renderer\Translate\Interceptor->Magento\Framework\Interception\{closure:vendor/magento/framework/Interception/Interceptor.php:104-151}()
PHP  57. Wirecard\ElasticEngine\Model\TranslationFallback\Interceptor->aroundRender()
PHP  58. Magento\Framework\Translate->loadData()
PHP  59. Magento\Framework\Translate->setConfig()
PHP  60. Magento\Theme\Model\View\Design\Proxy->getDesignTheme()
PHP  61. Magento\Theme\Model\View\Design->getDesignTheme()
PHP  62. Magento\Theme\Model\ThemeFactory->create()
PHP  63. Magento\Framework\App\ObjectManager->create()
PHP  64. Magento\Framework\ObjectManager\Factory\Compiled->create()
PHP  65. Magento\Framework\ObjectManager\Factory\Compiled->get()
PHP  66. Magento\Framework\ObjectManager\Factory\Compiled->create()
PHP  67. Magento\Framework\View\Design\Theme\Validator->__construct()
PHP  68. Magento\Framework\View\Design\Theme\Validator->_setTypeValidators()
PHP  69. Magento\Framework\Phrase->__toString()
PHP  70. Magento\Framework\Phrase->render()
PHP  71. Magento\Framework\Phrase\Renderer\Composite->render()
PHP  72. Magento\Framework\Phrase\Renderer\Translate\Interceptor->render()
PHP  73. Magento\Framework\Phrase\Renderer\Translate\Interceptor->___callPlugins()
PHP  74. Magento\Framework\Phrase\Renderer\Translate\Interceptor->Magento\Framework\Interception\{closure:vendor/magento/framework/Interception/Interceptor.php:104-151}()
PHP  75. Wirecard\ElasticEngine\Model\TranslationFallback\Interceptor->aroundRender()
PHP  76. Magento\Framework\Translate->loadData()
PHP  77. Magento\Framework\Translate->setConfig()
PHP  78. Magento\Theme\Model\View\Design\Proxy->getDesignTheme()
PHP  79. Magento\Theme\Model\View\Design->getDesignTheme()
PHP  80. Magento\Theme\Model\ThemeFactory->create()
PHP  81. Magento\Framework\App\ObjectManager->create()
PHP  82. Magento\Framework\ObjectManager\Factory\Compiled->create()
PHP  83. Magento\Framework\ObjectManager\Factory\Compiled->get()
PHP  84. Magento\Framework\ObjectManager\Factory\Compiled->create()
PHP  85. Magento\Framework\View\Design\Theme\Validator->__construct()
PHP  86. Magento\Framework\View\Design\Theme\Validator->_setTypeValidators()
PHP  87. Magento\Framework\Phrase->__toString()
PHP  88. Magento\Framework\Phrase->render()
PHP  89. Magento\Framework\Phrase\Renderer\Composite->render()
PHP  90. Magento\Framework\Phrase\Renderer\Translate\Interceptor->render()
PHP  91. Magento\Framework\Phrase\Renderer\Translate\Interceptor->___callPlugins()
PHP  92. Magento\Framework\Phrase\Renderer\Translate\Interceptor->Magento\Framework\Interception\{closure:vendor/magento/framework/Interception/Interceptor.php:104-151}()
PHP  93. Wirecard\ElasticEngine\Model\TranslationFallback\Interceptor->aroundRender()
PHP  94. Magento\Framework\Translate->loadData()
PHP  95. Magento\Framework\Translate->setConfig()
PHP  96. Magento\Theme\Model\View\Design\Proxy->getDesignTheme()
PHP  97. Magento\Theme\Model\View\Design->getDesignTheme()
PHP  98. Magento\Theme\Model\ThemeFactory->create()
PHP  99. Magento\Framework\App\ObjectManager->create()
PHP 100. Magento\Framework\ObjectManager\Factory\Compiled->create()
PHP 101. Magento\Framework\ObjectManager\Factory\Compiled->get()
PHP 102. Magento\Framework\ObjectManager\Factory\Compiled->create()
PHP 103. Magento\Framework\View\Design\Theme\Validator->__construct()
PHP 104. Magento\Framework\View\Design\Theme\Validator->_setTypeValidators()
PHP 105. Magento\Framework\Phrase->__toString()
PHP 106. Magento\Framework\Phrase->render()
PHP 107. Magento\Framework\Phrase\Renderer\Composite->render()
PHP 108. Magento\Framework\Phrase\Renderer\Translate\Interceptor->render()
PHP 109. Magento\Framework\Phrase\Renderer\Translate\Interceptor->___callPlugins()
PHP 110. Magento\Framework\Phrase\Renderer\Translate\Interceptor->Magento\Framework\Interception\{closure:vendor/magento/framework/Interception/Interceptor.php:104-151}()
```

**Steps to reproduce:**

Currently no public reproducer, since ours depends on proprietary code.

**Comment by rinnhofer** [comment]
Hello @nussjustin-hmmh ,

thank you for your contribution and the detailed explanation. We will look into it and try to reproduce your issue within our system + find a possible solution.

Kind regards,
Wirecard Shop Systems Team

**Comment by cpfoser** [comment]
Hello @nussjustin-hmmh,

unfortunately that issue depends on Magento2, because they does not support a fallback translation machanism for i18n. 

We will report this issue to Magento 2 directly. Further we will map en_US to languages which are not supported by us, to ensure that English is displayed instead of the key. 

Kind regards, 
Wirecard Shop Systems Team

**Comment by nussjustin-hmmh** [comment]
The change in #293 work around this bug by not loading the translations recursively. Fixing the fallback logic itself would be more complicated, since this module does translations and translation fallback in a non-standard way which I have never seen in any other Magento (1 or 2) module instead of using the standard way (using English words or sentences as translations keys).

I also just created a small module that reproduces the problem. The module can be found here: https://github.com/nussjustin-hmmh/Hmmh_WirecardExample

To reproduce the bug:
* Install the wirecard/magento2-ee module
* Checkout my repository
* Copy the files to `app/code/Hmmh/WirecardExample`
* Run `bin/magento setup:di:compile` and `bin/magento cache:flush`
* Finally just run `bin/magento hmmh:wirecardexample:example`

With Xdebug enabled this should output something like:

```
PHP Fatal error:  Method Magento\Framework\Phrase::__toString() must not throw an exception, caught Error: Maximum function nesting level of '256' reached, aborting! in vendor/magento/framework/View/Design/Theme/Validator.php on line 0
PHP Stack trace:
PHP   1. {main}() bin/magento:0
PHP   2. Magento\Framework\Console\Cli->run() bin/magento:23
PHP   3. Magento\Framework\Console\Cli->doRun() vendor/symfony/console/Application.php:145
PHP   4. Magento\Framework\Console\Cli->doRun() vendor/magento/framework/Console/Cli.php:102
PHP   5. Magento\Framework\Console\Cli->doRunCommand() vendor/symfony/console/Application.php:262
PHP   6. Hmmh\WirecardExample\Console\Command\ExampleCommand->run() vendor/symfony/console/Application.php:893
PHP   7. Hmmh\WirecardExample\Console\Command\ExampleCommand->execute() vendor/symfony/console/Command/Command.php:255
PHP   8. Magento\Framework\App\ObjectManager->create() app/code/Hmmh/WirecardExample/Console/Command/ExampleCommand.php:60
PHP   9. Magento\Framework\ObjectManager\Factory\Compiled->create() vendor/magento/framework/ObjectManager/ObjectManager.php:56
PHP  10. Magento\Framework\View\Design\Theme\Validator->__construct() vendor/magento/framework/ObjectManager/Factory/Compiled.php:60
PHP  11. Magento\Framework\View\Design\Theme\Validator->_setTypeValidators() vendor/magento/framework/View/Design/Theme/Validator.php:34
```

Without Xdebug this will lead to either an OOM error (in case of hitting the PHP `memory_limit`):

```
PHP Fatal error:  Allowed memory size of 2147483648 bytes exhausted (tried to allocate 20480 bytes) in vendor/magento/framework/App/State.php on line 153

Check https://getcomposer.org/doc/articles/troubleshooting.md#memory-limit-errors for more info on how to handle out of memory errors.
```

or (in case of OOM at the OS level)

```
mmap() failed: [12] Cannot allocate memory

mmap() failed: [12] Cannot allocate memory
PHP Fatal error:  Out of memory (allocated 5501091840) (tried to allocate 20480 bytes) in vendor/magento/framework/App/State.php on line 153
```

or otherwise a SEGFAULT.

The reproducer can be summarized as:

```php
\Magento\Framework\Phrase::setRenderer($di->create('Magento\Framework\Phrase\Renderer\Translate'));
/** @var \Magento\Framework\ObjectManagerInterface $di */
$di->create('Magento\Framework\View\Design\Theme\Validator')
```

You can also paste this snippet directly into a console when using the `dev:console` command from [n98-magerun2](https://github.com/netz98/n98-magerun2):

```php
$ n98-magerun2.phar dev:cons
At the prompt, type help for some help.

To exit the shell, type ^D.
Psy Shell v0.9.9 (PHP 7.2.17 — cli) by Justin Hileman
>>> \Magento\Framework\Phrase::setRenderer($di->create('Magento\Framework\Phrase\Renderer\Translate'));
=> null
>>> /** @var \Magento\Framework\ObjectManagerInterface $di */
>>> $di->create('Magento\Framework\View\Design\Theme\Validator')
PHP Fatal error:  Allowed memory size of 2147483648 bytes exhausted (tried to allocate 20480 bytes) in vendor/magento/framework/App/State.php on line 153

Check https://getcomposer.org/doc/articles/troubleshooting.md#memory-limit-errors for more info on how to handle out of memory errors.
```

**Comment by rinnhofer** [comment]
Hello @nussjustin-hmmh and @MatthiasHarteAtHMMH ,

we included the fix within #293 in our latest release https://github.com/wirecard/magento2-ee/releases/tag/2.2.1

Thank you @nussjustin-hmmh for the detailed explanation and reproduction steps and thank you @MatthiasHarteAtHMMH for your pull request.

We will continue working on this topic in the future and try to improve our translation process within our extension based on your input. For now I will close this issue.

Kind regards,
Wirecard Shop Systems Team
